src/vision/status_quo/alan_iteratively_regresses.md

```diff
@@ -0,0 +1,90 @@
+
+## The story
+
+A core part of DistriData, called DDSplit, is in charge of splitting input data records into fragments that are stored on distinct servers, and then reassembling those fragments back into records in response to user queries.
+
+DDSplit was originally implemented using Java code (plus some C, interfaced via JNI). Alan thinks that Rust could provide the same quality of service while requiring less memory, and decides to try reimplementing DDSplit in Rust.
+
+Alan develops a prototype atop tokio and the [async-trait crate][], leveraging `#[async_trait]`to be able to define `async fn` methods within traits. When Alan finishes the prototype code, he finds the prototype performance has 20% slower throughput compared to the Java version.
+
+[async-trait crate]: https://crates.io/crates/async-trait
+[async-trait transform]: https://crates.io/crates/async-trait#explanation
+
+Alan is disappointed; his experience has been that Rust code performs great, once you managed to get the code to be accepted by the compiler. Alan was not expecting to suffer a 20% performance hit over the Java code.
+
+The DDSplit service is being developed on a Linux machine, so Alan is able use the `perf` tool to gather sampling-based profiling data the async/await port of DDSplit. 
+
+Looking at a [flamegraph][] for the call stacks, Alan  identified many calls into the memory allocator, and many calls to `memcpy`.
+
+[flamegraph]: https://crates.io/crates/flamegraph
+
+Alan reaches out to Barbara, as the local Rust expert, for help on how identify where the performance pitfalls are coming from.
+
+Alan asks Barbara whether the problem could be caused by the tokio executor. Barbara says it is hard to know that without more instrumentation. She explains it *could* be that the program is overloading tokio's task scheduler (for example), but it also could be that the application code itself has expensive operations, such as lots of small I/O operations rather than using a buffer.
+
+Alan and Barbara look at the `perf` data. They find the output of `perf report` difficult to navigate and interpret. The data has stack trace fragments available, which gives them a few hints to follow up on. But when they try to make `perf report` annotate the original source, `perf` only shows disassembled machine code, not the original Rust source code. Alan and Barbara both agree that trying to dissect the problem from the machine code is not an attractive strategy.
+
+For the memory allocator issues, Barbara recommnds that Alan try to eliminate the allocation calls and if they cannot be eliminated, then that Alan try tuning the parameters for the global memory allocator, or even switching which global memory allocator he is using. Alan looks at Barbara in despair: his time tweaking GC settings on the Java Virtual Machine taught him that allocator tuning is often a black art.
```

Comment:
```suggestion
For the memory allocator issues, Barbara recommends that Alan try to eliminate the allocation calls and if they cannot be eliminated, then that Alan try tuning the parameters for the global memory allocator, or even switching which global memory allocator he is using. Alan looks at Barbara in despair: his time tweaking GC settings on the Java Virtual Machine taught him that allocator tuning is often a black art.
```

Reply:
Thanks, fixed in 8b9d4e1284cebf4074f353390934a9368a27920e